bpf: Add option to utilize core maps via BPF_F_NO_COMMON_LRU

borkmann opened the pull request:
(see commit msg)

borkmann committed:
bpf: Add option to utilize core maps via BPF_F_NO_COMMON_LRU

Avoid hitting spinlock contention for htab_lru_map_update_elem when
bpf_lru_pop_free tries to grab the backend memory pool's spinlock in
kernel. Instead switch to BPF_F_NO_COMMON_LRU. This greatly improves
performance under DDoS-like traffic (e.g. it was seen by a user that
Pod traffic bumped from 400kpps to ~1.6m pps with the flag enabled).

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium, helm: Add bpf.distributedLRU.enabled

Add a Helm setting called bpf.distributedLRU.enabled.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium, docs: Update performance guide on LRU

Add a section on this topic.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium, docs: Update performance guide on clock probe

Also add a section of BPF clock source.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

tklauser reviewed (approved):
Looks good for my code owners.

Also checked that we don't need to make `BPF_F_NO_COMMON_LRU` use conditional on kernel version as it was introduced in 4.10 and Cilium's minimum required kernel version is 5.4.

qmonnet reviewed (approved):
Looks good, but would it make sense to enable the option in some of the tests, to detect regressions?

borkmann committed:
cilium, tests: Enable distributedLRU in netkit e2e tests

Enable distributedLRU in the netkit e2e tests in order to excercise
this code in CI as well.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

ysksuzuki reviewed (approved):
The changes look good for me.

There are some LRU hash maps that are not supported, such as IPV4_FRAG_DATAGRAMS_MAP and RATELIMIT_MAP. If they were intentionally excluded because they are not used frequently, then that's not a problem.
